python/strings_udf/cpp/src/strings/udf/shim.cu

```diff
@@ -270,3 +271,57 @@ extern "C" __device__ int rstrip(int* nb_retval,
 
   return 0;
 }
+
+extern "C" __device__ int upper(int* nb_retval,
+                                void* udf_str,
+                                void* const* st,
+                                std::int64_t flags_table,
+                                std::int64_t cases_table,
+                                std::int64_t special_table)
+{
+  auto udf_str_ptr = new (udf_str) udf_string;
+  auto st_ptr      = reinterpret_cast<cudf::string_view const*>(st);
+
+  auto flags_table_ptr =
+    reinterpret_cast<cudf::strings::detail::character_flags_table_type*>(flags_table);
+  auto cases_table_ptr =
+    reinterpret_cast<cudf::strings::detail::character_cases_table_type*>(cases_table);
+  auto special_table_ptr =
+    reinterpret_cast<cudf::strings::detail::special_case_mapping*>(special_table);
+
+  cudf::strings::udf::chars_tables tables;
+  tables.flags_table                = flags_table_ptr;
+  tables.cases_table                = cases_table_ptr;
+  tables.special_case_mapping_table = special_table_ptr;
+
+  *udf_str_ptr = to_upper(tables, *st_ptr);
+
+  return 0;
+}
+
+extern "C" __device__ int lower(int* nb_retval,
+                                void* udf_str,
+                                void* const* st,
```

Comment:
Does this need to be a `void **` or can it just be a `void *` (I note that inside you cast to `string_view *` and then dereference, so I think you can strip a `*` everywhere).

Reply:
I am not sure I follow here. This follows the pattern from the rest of the shim functions when a `string_view` is an arg. IIUC `st` is pointing directly to the struct, so only one level of pointing right?

Reply:
Most functions here (e.g. https://github.com/rapidsai/cudf/blob/418cbc69b336f34b06e1de21ae2c016b40f1e6fc/python/strings_udf/cpp/src/strings/udf/shim.cu#L186-L193) that take a point to `string_view` take the argument as `void const *`, not `void* const *`. Why is this different?